packages/sources/coinapi-test/src/endpoint/crypto-ws.ts

```diff
@@ -0,0 +1,61 @@
+import { WebSocketTransport } from '@chainlink/external-adapter-framework/transports/websocket'
+import { SingleNumberResultResponse } from '@chainlink/external-adapter-framework/util'
+import { customSettings } from '../config'
+import { RequestParams } from '../crypto-utils'
+
+export interface WsMessage {
+  time: string
+  asset_id_base: string
+  asset_id_quote: string
+  rate: number
+  type: string
+}
+
+export type EndpointTypes = {
+  Request: {
+    Params: RequestParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    WsMessage: WsMessage
+  }
+}
+
+let apikey = ''
+export const wsTransport = new WebSocketTransport<EndpointTypes>({
+  url: (context) => {
+    apikey = context.adapterConfig.API_KEY
+    return context.adapterConfig.WS_API_ENDPOINT
+  },
+  handlers: {
+    message(message) {
+      if (message.type !== 'exrate') {
+        return []
+      }
+      return [
+        {
+          params: { base: message.asset_id_base, quote: message.asset_id_quote },
+          response: {
+            data: {
+              result: message.rate,
+            },
+            result: message.rate,
+          },
+        },
+      ]
+    },
+  },
+
+  builders: {
+    subscribeMessage: (params) => {
+      return {
+        type: 'hello',
+        apikey: apikey,
+        heartbeat: false,
+        subscribe_data_type: ['exrate'],
+        subscribe_filter_asset_id: [`${params.base}/${params.quote}`],
```

Comment:
Looking at CoinAPI v2, I think it passes base and quote separately in the `subscribe_filter_asset_id` array. Was it changed to send a specific pair like this for a reason? Based on CoinAPI docs, it seems like this is valid too but unsure if there's any behavior difference

Reply:
Yes there is a behavior difference. In v2 when we pass two separate arguments coinapi treats them as two 'bases' and returns every possible quote information. For example for ['BTC' , 'ETH'] coinapi will return every price info for BTC/USD, BTC/DAO, BTC/SOL and so on, same for ETH. 
Whereas if we send ['BTC/ETH'] coinapi will return only prices for this particular pair. 
Not sure which one is better for us.

Reply:
Ok we can keep it how you have it now. Probably better we don't have process extra pairs that would be unused. Don't think we have too many pairs configured for the same base where this would help. And it would only help in response time for the initial request anyways.